src/algebra/lie/weights.lean

```diff
@@ -231,37 +231,142 @@ end lie_module
               ∀ (y : H), ∃ (k : ℕ), (H.incl : H →ₗ[R] L).comp ((g y)^k) x = 0,
   { simp only [lie_hom.coe_to_linear_map, lie_subalgebra.coe_incl, function.comp_app,
       linear_map.coe_comp, submodule.coe_eq_zero] at this,
-    simp only [lie_module.mem_weight_space, lie_module.mem_pre_weight_space,
+    simp only [mem_weight_space, mem_pre_weight_space,
       lie_subalgebra.coe_incl', lie_submodule.mem_comap, this], },
   have hfg : ∀ (y : H), (f y).comp (H.incl : H →ₗ[R] L) = (H.incl : H →ₗ[R] L).comp (g y),
   { rintros ⟨y, hz⟩, ext ⟨z, hz⟩,
-    simp only [submodule.coe_sub, lie_module.to_endomorphism_apply_apply, lie_hom.coe_to_linear_map,
+    simp only [submodule.coe_sub, to_endomorphism_apply_apply, lie_hom.coe_to_linear_map,
       linear_map.one_apply, lie_subalgebra.coe_incl, lie_subalgebra.coe_bracket_of_module,
       lie_subalgebra.coe_bracket, linear_map.smul_apply, function.comp_app,
       submodule.coe_smul_of_tower, linear_map.coe_comp, linear_map.sub_apply], },
   simp_rw [linear_map.commute_pow_left_of_commute (hfg _)],
 end
 
+variables {H M}
+
+lemma lie_mem_weight_space_of_mem_weight_space {χ₁ χ₂ : H → R} {x : L} {m : M}
+  (hx : x ∈ root_space H χ₁) (hm : m ∈ weight_space M χ₂) : ⁅x, m⁆ ∈ weight_space M (χ₁ + χ₂) :=
+begin
+  apply lie_module.weight_vector_multiplication L M M
+    (H.restrict_lie_module_hom (to_module_hom R L M)) χ₁ χ₂,
+  simp only [lie_module_hom.coe_to_linear_map, function.comp_app, linear_map.coe_comp,
+    tensor_product.map_incl, linear_map.mem_range],
+  use [⟨x, hx⟩ ⊗ₜ ⟨m, hm⟩],
+  simp only [submodule.subtype_apply, to_module_hom_apply, submodule.coe_mk,
+    lie_subalgebra.coe_restrict_lie_module_hom, tensor_product.map_tmul],
+end
+
+variables (R L H M)
+
+/-- Given a nilpotent Lie subalgebra `H ⊆ L` together with `χ₁ χ₂ : H → R`, there is a natural
+`R`-linear product of root vectors and weight vectors, compatible with the actions of `H`. -/
+def root_space_weight_space_product (χ₁ χ₂ : H → R) :
+  (root_space H χ₁) ⊗[R] (weight_space M χ₂) →ₗ⁅R,H⁆ weight_space M (χ₁ + χ₂) :=
+lift_lie R H (root_space H χ₁) (weight_space M χ₂) (weight_space M (χ₁ + χ₂))
+{ to_fun    := λ x,
+  { to_fun    :=
+      λ m, ⟨⁅(x : L), (m : M)⁆, lie_mem_weight_space_of_mem_weight_space x.property m.property⟩,
+    map_add'  := λ m n, by { simp only [lie_submodule.coe_add, lie_add], refl, },
+    map_smul' := λ t m, by { conv_lhs { congr, rw [lie_submodule.coe_smul, lie_smul], }, refl, }, },
+  map_add'  := λ x y, by ext m; rw [linear_map.add_apply, linear_map.coe_mk, linear_map.coe_mk,
+    linear_map.coe_mk, subtype.coe_mk, lie_submodule.coe_add, lie_submodule.coe_add, add_lie,
+    subtype.coe_mk, subtype.coe_mk],
+  map_smul' := λ t x, by ext m; rw [linear_map.smul_apply, linear_map.coe_mk, linear_map.coe_mk,
+    subtype.coe_mk, lie_submodule.coe_smul, smul_lie, lie_submodule.coe_smul, subtype.coe_mk],
+  map_lie'  := λ x y, by ext m; rw [lie_hom.lie_apply, lie_submodule.coe_sub, linear_map.coe_mk,
```

Comment:
For reasons I don't understand I must use `rw` in most of the proof obligations of this definition.

For example if I replace the `rw` here with `simp only` then I get a deterministic timeout (even with a very long time limit). In fact even taking just the first two terms of this `rw`, namely `simp only [lie_hom.lie_apply, lie_submodule.coe_sub]` times out (after first consuming more and more memory).

Is there a simp loop? I don't really know how to debug this.

Reply:
@kbuzzard  just pointed out to me the problems with `simp` are now gone on current master. Indeed the following works:
```lean
def root_space_weight_space_product (χ₁ χ₂ χ₃ : H → R) (hχ : χ₁ + χ₂ = χ₃) :
  (root_space H χ₁) ⊗[R] (weight_space M χ₂) →ₗ⁅R,H⁆ weight_space M χ₃ :=
lift_lie R H (root_space H χ₁) (weight_space M χ₂) (weight_space M χ₃)
{ to_fun    := λ x,
  { to_fun    :=
      λ m, ⟨⁅(x : L), (m : M)⁆,
            hχ ▸ (lie_mem_weight_space_of_mem_weight_space x.property m.property) ⟩,
    map_add'  := λ m n, by simpa,
    map_smul' := λ t m, by simpa, },
  map_add'  := λ x y, by ext m; simp,
  map_smul' := λ t x, by ext m; simp,
  map_lie'  := λ x y, by ext m; simp, }
```
However I will leave the implementation as it stands because elaboration time on my machine increases from about 3.4s to 12.8s if I switch to the above.

It is a mystery as to what the problem was but I can reproduce it by navigating to the commit #edf56dfce079f0732084fda74073ecd6c19229d2 on this (temporarily restored) branch.

Using `set_option trace.simplify.rewrite true` I see now loop and using the more verbose `set_option trace.simplify true`, I see nothing obvious, but I do not understand its output well.

As regards why this works on master but not on the old commit, I tested and it appears to be the result of the changed definition from:
```lean
def root_space_weight_space_product (χ₁ χ₂ : H → R) :
```
to:
```
def root_space_weight_space_product (χ₁ χ₂ χ₃ : H → R) (hχ : χ₁ + χ₂ = χ₃) :
```
that was added (for other reasons) in the final commit of this PR.